### DISCUSSION TITLE
Cache model functions for iterative workflows

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pymc3#7177] [Ideas]
Users are often mislead to believe that when defining models with MutableData, PyMC will avoid recompiling logp/prior/posterior predictive functions when only the MutableData has changed. This is not the case! 

PyMC always recompiles functions when you call `pm.sample` or `pm.sample_posterior_predictive`. We used to cache some of the model methods but this was removed because models are not static objects and new variables can be added any time: https://github.com/pymc-devs/pymc/pull/2345/commits/983d44430b900680685e90ccfe3de347f846f9d6#diff-f593a76ecc6f9a5c5abdd7bbd3e9de9add74a068e75e64f66b7b1424a279a0dd

There's a couple approaches we could take to give back to users the caching functionality:
1. Allow caching, but delete all cache whenever a model is mutated, which should happen only when `model.register_rv` is called or `model.rvs_to_transforms` is changed. We could easily forbid the latter now that we have https://www.pymc.io/projects/docs/en/v5.10.3/api/model/generated/pymc.model.transform.conditioning.change_value_transforms.html
2. Provide a function to "freeze" a PyMC model. Then the model, much like nutpie `compiled_pymc` can never be mutated and caching would be safe.
3. Provide a better low-level API for users to request model functions and use them in sampling routines

We may need to include the contents of `pytensor.config` as part of the caching / freezing so that users can request the same functions in different backends.

## Linked issues:
* https://github.com/pymc-devs/pymc/discussions/6239
* https://github.com/pymc-devs/pymc/issues/7073

**Comment by twiecki** [comment]
I like the options in this order: 1, 2, 3.

**Reply by ricardoV94** [reply]
Me too, only thing is that caching has a tendency to turn out more complex that it seems at first glance. I do think this will be less of an issue as we keep moving into a more functional workflow where we don't modify models inplace but instead recreate new ones like we do in `change_value_transforms`, `do`, `observed`

**Comment by zaxtax** [comment]
I'm a fan of option 1.

**Comment by twiecki** [comment]
One thing that would be nice for deployment is if I could pre-compile a model so that when a user wants to run something it doesn't have to compile at that point.

**Reply by ricardoV94** [reply]
That's a completely different challenge level, requires ahead of time compilation. I don't think any of the backends has anything good there/yet
